probe_scraper/transform_pings.py

```diff
@@ -0,0 +1,140 @@
+    if ping in repo_pings:
+        prev_defns = repo_pings[ping][HISTORY_KEY]
+        max_defn_i = max(range(len(prev_defns)),
+                         key=lambda i: datetime.fromisoformat(prev_defns[i][DATES_KEY]["last"]))
+        max_defn = prev_defns[max_defn_i]
+
+        # If equal to previous commit, update date and commit on existing definition
+        if pings_equal(definition, max_defn):
+            new_defn = make_ping_defn(max_defn, commit_hash, commit_timestamps)
+            repo_pings[ping][HISTORY_KEY][max_defn_i] = new_defn
+
+        # Otherwise, prepend changed definition for existing ping
+        else:
+            new_defn = make_ping_defn(definition, commit_hash, commit_timestamps)
+            repo_pings[ping][HISTORY_KEY] = prev_defns + [new_defn]
+
+    # We haven't seen this ping before, add it
+    else:
+        defn = make_ping_defn(definition, commit_hash, commit_timestamps)
+        repo_pings[ping] = {
+            NAME_KEY: ping,
+            HISTORY_KEY: [defn]
+        }
+
+    return repo_pings
+
+
+def transform_by_hash(commit_timestamps, ping_data):
+    """
+    :param commit_timestamps - of the form
+      <repo_name>: {
+        <commit-hash>: <commit-timestamp>,
+        ...
+      }
+
+    :param ping_data - of the form
+      <repo_name>: {
+        <commit-hash>: {
+          <ping-name>: {
+            ...
+          },
+        },
+        ...
+      }
+
+    Outputs deduplicated data of the form
+        <repo_name>: {
+            <ping_name>: {
+                "type": <type>,
+                "name": <name>,
+                "history": [
+                    {
+                        "bugs": [<bug#>, ...],
+                        ...other pings.yaml info...,
+                        "git-commits": {
+                            "first": <hash>,
+                            "last": <hash>
+                        },
+                        "dates": {
+                            "first": <datetime>,
+                            "last": <datetime>
+                        }
+                    },
+                ]
+            }
+        }
+    """
+
+    all_pings = {}
+    for repo_name, commits in ping_data.items():
+        repo_pings = {}
+
+        # iterate through commits, sorted by timestamp of the commit
+        sorted_commits = sorted(iter(commits.items()),
+                                key=lambda x_y: commit_timestamps[repo_name][x_y[0]])
+
+        for commit_hash, pings in sorted_commits:
+            for ping, definition in pings.items():
+                repo_pings = update_or_add_ping(repo_pings,
+                                                commit_hash,
+                                                ping,
+                                                definition,
+                                                commit_timestamps[repo_name])
+
+        all_pings[repo_name] = repo_pings
+
+    return all_pings
```

Comment:
This seems really redundant with transform_metrics.py.  Maybe we can just put the pings as in the `pings` category in the metrics data structure (this is what glean_parser does).  Then it can all be handled by the same code...

Reply:
Unfortunately the transform_probes part hard-codes the `TYPE_KEY`, which pings don't have. 
It might be easier to just refactor out that small part and otherwise reuse the code.

Reply:
:+1: